aiohttp/http_websocket.py

```diff
@@ -626,28 +627,48 @@ async def _send_frame(
         if (compress or self.compress) and opcode < 8:
             if compress:
                 # Do not set self._compress if compressing is for this frame
-                compressobj = ZLibCompressor(
-                    level=zlib.Z_BEST_SPEED,
-                    wbits=-compress,
-                    max_sync_chunk_size=WEBSOCKET_MAX_SYNC_CHUNK_SIZE,
-                )
+                compressobj = self._make_compress_obj(compress)
+                message = await compressobj.compress(message)
+                self._write_compressed_message(message, opcode, compressobj, rsv)
             else:  # self.compress
+                # If we are reusing the compress object, we need to hold a lock
+                # around the compress/flush/write to ensure that messages get
+                # compressed in order and takeover is not violated.
                 if not self._compressobj:
-                    self._compressobj = ZLibCompressor(
-                        level=zlib.Z_BEST_SPEED,
-                        wbits=-self.compress,
-                        max_sync_chunk_size=WEBSOCKET_MAX_SYNC_CHUNK_SIZE,
-                    )
+                    self._compressobj = self._make_compress_obj(self.compress)
+                    self._compress_lock = asyncio.Lock()
                 compressobj = self._compressobj
+                async with self._compress_lock:
```

Comment:
I'm just thinking about this again. Would it be better to have the lock in the Compressor, so the logic is encapsulated and not going to cause issues if used elsewhere?

From my reading, the only yield point of this logic is when it's offloaded to a thread. If there is a lock around that call, then no other coro can enter it until it resumes. Once it resumes it will continue to call the remaining functions without yielding, so it should be safe?

Reply:
I _think_ that is safe since releasing the lock will only [wake up the next waiter](https://github.com/python/cpython/blob/dc0adb44d8d4a33121deaad398f24b5d8ae36d19/Lib/asyncio/locks.py#L137C18-L137C32) and the original should proceed on to flush and writer since [`set_result`](https://github.com/python/cpython/blob/dc0adb44d8d4a33121deaad398f24b5d8ae36d19/Lib/asyncio/locks.py#L154) will result in a [`call_soon`](https://github.com/python/cpython/blob/dc0adb44d8d4a33121deaad398f24b5d8ae36d19/Lib/asyncio/futures.py#L173) from [`__schedule_callbacks`](https://github.com/python/cpython/blob/dc0adb44d8d4a33121deaad398f24b5d8ae36d19/Lib/asyncio/futures.py#L263).  

That might be a implementation detail of the event loop implementation rather than a guarantee though

Reply:
OTOH, I think the only place we need a lock for compression is the websocket and the other use cases are already fine

```
aiohttp/http_writer.py:            chunk = await self._compress.compress(chunk)
aiohttp/http_writer.py:                chunk = await self._compress.compress(chunk)
aiohttp/multipart.py:                chunk = await self._compress.compress(chunk)
```

Reply:
I don't think the event loop implementation matters. As long as the lock's release doesn't yield to the event loop (which would be an odd thing to do, I think), then there is no opportunity for another coro to run before we complete the other code.

Reply:
> OTOH, I think the only place we need a lock for compression is the websocket and the other use cases are already fine

Is it?

> ```
> aiohttp/http_writer.py:            chunk = await self._compress.compress(chunk)
> aiohttp/http_writer.py:                chunk = await self._compress.compress(chunk)
> aiohttp/multipart.py:                chunk = await self._compress.compress(chunk)
> ```

While I think it's slightly less likely you'd attempt to make multiple StreamResponse.write() calls than on a websocket, it's still possible and looks to me like it'd have the same issue? Maybe the same for multipart..

Reply:
e.g. I'm making some kind of custom output, where I'm just writing chunks out to the response, as those chunks become available asynchronously.

Reply:
I think most of the decompress ones are fine, but only due to implementation details. i.e. They seem to be in feed_data() implementations, which I assume are called in a loop or similar as the parser goes through the data.

I think the one in multipart could have a tiny chance of causing issues, but it's been left as a sync method anyway, so will be fine.

Reply:
Assuming you are confident it's safe I can refactor this after the holiday to move the lock into the compressor class

Reply:
I'm pretty sure it's safe today. The only question is how might the code be refactored in future. If someone adds an await between the calls, then it'll be unsafe. But, if we don't change, and then someone uses the compressor somewhere else without using a lock, then that'll be unsafe. So, there's a risk either way. To me, keeping it encapsulated seems slightly less worse. :P

Reply:
Haha. I think that's a yes on the refactor. I'll work on it after all the family events

Reply:
I refactored it and added some comments in the hope that it will help avoid problems from potential future refactoring you mention above.